File: src/SnapBrief.Core/Editing/CaptureCropper.cs
using System.Collections.Immutable;
using SnapBrief.Core.Models;

namespace SnapBrief.Core.Editing;

public readonly record struct NormalizedRect(double X, double Y, double Width, double Height)
{
    public double Left => X;
    public double Top => Y;
    public double Right => X + Width;
    public double Bottom => Y + Height;
}

public sealed record CropResult(
    CaptureItem PreviousCapture,
    CaptureItem CroppedCapture,
    ImmutableArray<Guid> RemovedAnnotationIds);

public static class CaptureCropper
{
    private const double Epsilon = 1e-12;

    public static CropResult Crop(
        CaptureItem source,
        NormalizedRect cropBounds,
        string croppedSourceImagePath,
        int croppedPixelWidth,
        int croppedPixelHeight)
    {
        ArgumentNullException.ThrowIfNull(source);
        ValidateCrop(cropBounds, croppedSourceImagePath, croppedPixelWidth, croppedPixelHeight);

        var retained = ImmutableArray.CreateBuilder<AnnotationItem>();
        var removed = ImmutableArray.CreateBuilder<Guid>();
        foreach (var annotation in source.Annotations)
        {
            if (annotation.Kind is AnnotationKind.Freehand or AnnotationKind.Highlight)
            {
                var croppedSegments = CropPolyline(annotation.GetPathSegments(), cropBounds);
                if (croppedSegments.IsEmpty)
                {
                    removed.Add(annotation.Id);
                }
                else
                {
                    retained.Add(annotation with
                    {
                        Points = croppedSegments[0],
                        PathSegments = croppedSegments.Length > 1 ? croppedSegments : []
                    });
                }
                continue;
            }

            var croppedPoints = annotation.Kind switch
            {
                AnnotationKind.Arrow => CropLine(annotation.Points, cropBounds),
                AnnotationKind.Rectangle or AnnotationKind.Text or AnnotationKind.Redaction or AnnotationKind.Blur or AnnotationKind.Comment =>
                    CropBox(annotation.Points, cropBounds),
                _ => ImmutableArray<NormalizedPoint>.Empty
            };

            if (croppedPoints.IsEmpty)
            {
                removed.Add(annotation.Id);
            }
            else
            {
                retained.Add(annotation with { Points = croppedPoints });
            }
        }

        var retainedIds = retained.Select(annotation => annotation.Id).ToHashSet();
        for (var i = 0; i < retained.Count; i++)
            if (retained[i].ParentAnnotationId is { } parentId && !retainedIds.Contains(parentId))
                retained[i] = retained[i] with { ParentAnnotationId = null };

        // The badge offset is normalized against the image, and the crop makes the image smaller:
        // the same shift in pixels is a bigger fraction of the cropped picture.
        for (var i = 0; i < retained.Count; i++)
            if (retained[i].NoteOffset is { } offset)
                retained[i] = retained[i] with
                {
                    NoteOffset = new NormalizedPoint(offset.X / cropBounds.Width, offset.Y / cropBounds.Height)
                };

        var cropped = source with
        {
            SourceImagePath = croppedSourceImagePath,
            PixelWidth = croppedPixelWidth,
            PixelHeight = croppedPixelHeight,
            Annotations = retained.ToImmutable()
        };
        return new CropResult(source, cropped, removed.ToImmutable());
    }

    private static ImmutableArray<NormalizedPoint> CropLine(
        ImmutableArray<NormalizedPoint> points,
        NormalizedRect crop)
    {
        if (points.Length < 2 || !TryClipSegment(points[0], points[1], crop, out var start, out var end) ||
            Distance(start, end) <= Epsilon)
        {
            return ImmutableArray<NormalizedPoint>.Empty;
        }

        return [MapToCrop(start, crop), MapToCrop(end, crop)];
    }

    private static ImmutableArray<NormalizedPoint> CropBox(
        ImmutableArray<NormalizedPoint> points,
        NormalizedRect crop)
    {
        if (points.Length < 2)
        {
            return ImmutableArray<NormalizedPoint>.Empty;
        }

        var left = Math.Max(Math.Min(points[0].X, points[1].X), crop.Left);
        var top = Math.Max(Math.Min(points[0].Y, points[1].Y), crop.Top);
        var right = Math.Min(Math.Max(points[0].X, points[1].X), crop.Right);
        var bottom = Math.Min(Math.Max(points[0].Y, points[1].Y), crop.Bottom);
        if (right - left <= Epsilon || bottom - top <= Epsilon)
        {
            return ImmutableArray<NormalizedPoint>.Empty;
        }

        return [MapToCrop(new(left, top), crop), MapToCrop(new(right, bottom), crop)];
    }

    private static ImmutableArray<ImmutableArray<NormalizedPoint>> CropPolyline(
        ImmutableArray<ImmutableArray<NormalizedPoint>> paths,
        NormalizedRect crop)
    {
        var runs = new List<List<NormalizedPoint>>();
        foreach (var points in paths)
        {
            List<NormalizedPoint>? current = null;
            for (var index = 1; index < points.Length; index++)
            {
                if (!TryClipSegment(points[index - 1], points[index], crop, out var start, out var end) ||
                    Distance(start, end) <= Epsilon)
                {
                    FinishRun(ref current, runs);
                    continue;
                }

                if (current is null || Distance(current[^1], start) > Epsilon)
                {
                    FinishRun(ref current, runs);
                    current = [start];
                }
                if (Distance(current[^1], end) > Epsilon)
                {
                    current.Add(end);
                }
            }
            FinishRun(ref current, runs);
        }

        return runs
            .Where(run => run.Count >= 2)
            .Where(run => PolylineLength(run) > Epsilon)
            .Select(run => run.Select(point => MapToCrop(point, crop)).ToImmutableArray())
            .ToImmutableArray();
    }

    private static void FinishRun(ref List<NormalizedPoint>? current, List<List<NormalizedPoint>> runs)
    {
        if (current is { Count: >= 2 })
        {
            runs.Add(current);
        }
        current = null;
    }

    private static bool TryClipSegment(
        NormalizedPoint start,
        NormalizedPoint end,
        NormalizedRect crop,
        out NormalizedPoint clippedStart,
        out NormalizedPoint clippedEnd)
    {
        var deltaX = end.X - start.X;
        var deltaY = end.Y - start.Y;
        var minimum = 0d;
        var maximum = 1d;

        if (!ClipTest(-deltaX, start.X - crop.Left, ref minimum, ref maximum) ||
            !ClipTest(deltaX, crop.Right - start.X, ref minimum, ref maximum) ||
            !ClipTest(-deltaY, start.Y - crop.Top, ref minimum, ref maximum) ||
            !ClipTest(deltaY, crop.Bottom - start.Y, ref minimum, ref maximum))
        {
            clippedStart = default;
            clippedEnd = default;
            return false;
        }

        clippedStart = new(start.X + minimum * deltaX, start.Y + minimum * deltaY);
        clippedEnd = new(start.X + maximum * deltaX, start.Y + maximum * deltaY);
        return true;
    }

    private static bool ClipTest(double denominator, double numerator, ref double minimum, ref double maximum)
    {
        if (Math.Abs(denominator) <= Epsilon)
        {
            return numerator >= 0;
        }

        var ratio = numerator / denominator;
        if (denominator < 0)
        {
            if (ratio > maximum) return false;
            if (ratio > minimum) minimum = ratio;
        }
        else
        {
            if (ratio < minimum) return false;
            if (ratio < maximum) maximum = ratio;
        }
        return true;
    }

    private static NormalizedPoint MapToCrop(NormalizedPoint point, NormalizedRect crop) => new(
        Math.Clamp((point.X - crop.Left) / crop.Width, 0, 1),
        Math.Clamp((point.Y - crop.Top) / crop.Height, 0, 1));

    private static double PolylineLength(IReadOnlyList<NormalizedPoint> points)
    {
        var length = 0d;
        for (var index = 1; index < points.Count; index++)
        {
            length += Distance(points[index - 1], points[index]);
        }
        return length;
    }

    private static double Distance(NormalizedPoint first, NormalizedPoint second)
    {
        var x = second.X - first.X;
        var y = second.Y - first.Y;
        return Math.Sqrt(x * x + y * y);
    }

    private static void ValidateCrop(NormalizedRect crop, string sourcePath, int pixelWidth, int pixelHeight)
    {
        if (!double.IsFinite(crop.X) || !double.IsFinite(crop.Y) ||
            !double.IsFinite(crop.Width) || !double.IsFinite(crop.Height) ||
            crop.X < 0 || crop.Y < 0 || crop.Width <= 0 || crop.Height <= 0 ||
            crop.Right > 1 || crop.Bottom > 1)
        {
            throw new ArgumentOutOfRangeException(nameof(crop), "Crop bounds must be a positive rectangle within normalized [0, 1] image space.");
        }

        ArgumentException.ThrowIfNullOrWhiteSpace(sourcePath);
        if (Path.IsPathRooted(sourcePath) || sourcePath.Split(Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar).Contains(".."))
        {
            throw new ArgumentException("Cropped source image path must remain relative to the session directory.", nameof(sourcePath));
        }
        if (pixelWidth <= 0) throw new ArgumentOutOfRangeException(nameof(pixelWidth));
        if (pixelHeight <= 0) throw new ArgumentOutOfRangeException(nameof(pixelHeight));
    }
}

File: macos/Sources/SnapikCore/Editing/CaptureCropper.swift
import Foundation

/// Port of `NormalizedRect(double X, double Y, double Width, double Height)`
/// from `src/Snapik.Core/Editing/CaptureCropper.cs`.
public struct NormalizedRect: Equatable, Sendable {
    public var x: Double
    public var y: Double
    public var width: Double
    public var height: Double

    public init(_ x: Double, _ y: Double, _ width: Double, _ height: Double) {
        self.x = x
        self.y = y
        self.width = width
        self.height = height
    }

    public var left: Double { x }
    public var top: Double { y }
    public var right: Double { x + width }
    public var bottom: Double { y + height }
}

/// Port of `CropResult`.
public struct CropResult: Sendable {
    public let previousCapture: CaptureItem
    public let croppedCapture: CaptureItem
    public let removedAnnotationIds: [SBGuid]
}

/// Port of `src/Snapik.Core/Editing/CaptureCropper.cs`.
public enum CaptureCropper {
    private static let epsilon = 1e-12

    public static func crop(
        source: CaptureItem,
        cropBounds: NormalizedRect,
        croppedSourceImagePath: String,
        croppedPixelWidth: Int,
        croppedPixelHeight: Int
    ) throws -> CropResult {
        try validateCrop(
            cropBounds,
            sourcePath: croppedSourceImagePath,
            pixelWidth: croppedPixelWidth,
            pixelHeight: croppedPixelHeight)

        var retained: [AnnotationItem] = []
        var removed: [SBGuid] = []

        for annotation in source.annotations {
            if annotation.kind == .freehand || annotation.kind == .highlight {
                let croppedSegments = cropPolyline(annotation.getPathSegments(), cropBounds)
                if croppedSegments.isEmpty {
                    removed.append(annotation.id)
                } else {
                    var updated = annotation
                    updated.points = croppedSegments[0]
                    updated.pathSegments = croppedSegments.count > 1 ? croppedSegments : []
                    retained.append(updated)
                }
                continue
            }

            let croppedPoints: [NormalizedPoint]
            switch annotation.kind {
            case .arrow:
                croppedPoints = cropLine(annotation.points, cropBounds)
            case .rectangle, .text, .redaction, .blur, .comment:
                croppedPoints = cropBox(annotation.points, cropBounds)
            case .highlight, .freehand:
                croppedPoints = []
            }

            if croppedPoints.isEmpty {
                removed.append(annotation.id)
            } else {
                var updated = annotation
                updated.points = croppedPoints
                retained.append(updated)
            }
        }

        // Port of `:70-74` (SPEC-DELTA-2B §B): a linked comment whose parent annotation was
        // removed by the crop loses the link but keeps its own note/geometry.
        let retainedIds = Set(retained.map(\.id))
        for index in retained.indices where retained[index].parentAnnotationId.map({ !retainedIds.contains($0) }) == true {
            retained[index].parentAnnotationId = nil
        }

        var cropped = source
        cropped.sourceImagePath = croppedSourceImagePath
        cropped.pixelWidth = croppedPixelWidth
        cropped.pixelHeight = croppedPixelHeight
        cropped.annotations = retained

        return CropResult(previousCapture: source, croppedCapture: cropped, removedAnnotationIds: removed)
    }

    private static func cropLine(_ points: [NormalizedPoint], _ crop: NormalizedRect) -> [NormalizedPoint] {
        guard points.count >= 2, let clipped = tryClipSegment(points[0], points[1], crop),
            distance(clipped.0, clipped.1) > epsilon
        else {
            return []
        }
        return [mapToCrop(clipped.0, crop), mapToCrop(clipped.1, crop)]
    }

    private static func cropBox(_ points: [NormalizedPoint], _ crop: NormalizedRect) -> [NormalizedPoint] {
        guard points.count >= 2 else { return [] }

        let left = max(min(points[0].x, points[1].x), crop.left)
        let top = max(min(points[0].y, points[1].y), crop.top)
        let right = min(max(points[0].x, points[1].x), crop.right)
        let bottom = min(max(points[0].y, points[1].y), crop.bottom)
        if right - left <= epsilon || bottom - top <= epsilon {
            return []
        }

        return [
            mapToCrop(NormalizedPoint(left, top), crop),
            mapToCrop(NormalizedPoint(right, bottom), crop),
        ]
    }

    private static func cropPolyline(
        _ paths: [[NormalizedPoint]],
        _ crop: NormalizedRect
    ) -> [[NormalizedPoint]] {
        var runs: [[NormalizedPoint]] = []

        for points in paths {
            var current: [NormalizedPoint]?
            for index in stride(from: 1, to: points.count, by: 1) {
                guard let clipped = tryClipSegment(points[index - 1], points[index], crop),
                    distance(clipped.0, clipped.1) > epsilon
                else {
                    finishRun(&current, &runs)
                    continue
                }
                let (start, end) = clipped

                if current == nil || distance(current!.last!, start) > epsilon {
                    finishRun(&current, &runs)
                    current = [start]
                }
                if distance(current!.last!, end) > epsilon {
                    current!.append(end)
                }
            }
            finishRun(&current, &runs)
        }

        return runs
            .filter { $0.count >= 2 }
            .filter { polylineLength($0) > epsilon }
            .map { run in run.map { mapToCrop($0, crop) } }
    }

    private static func finishRun(_ current: inout [NormalizedPoint]?, _ runs: inout [[NormalizedPoint]]) {
        if let run = current, run.count >= 2 {
            runs.append(run)
        }
        current = nil
    }

    private static func tryClipSegment(
        _ start: NormalizedPoint,
        _ end: NormalizedPoint,
        _ crop: NormalizedRect
    ) -> (NormalizedPoint, NormalizedPoint)? {
        let deltaX = end.x - start.x
        let deltaY = end.y - start.y
        var minimum = 0.0
        var maximum = 1.0

        guard clipTest(-deltaX, start.x - crop.left, &minimum, &maximum),
            clipTest(deltaX, crop.right - start.x, &minimum, &maximum),
            clipTest(-deltaY, start.y - crop.top, &minimum, &maximum),
            clipTest(deltaY, crop.bottom - start.y, &minimum, &maximum)
        else {
            return nil
        }

        let clippedStart = NormalizedPoint(start.x + minimum * deltaX, start.y + minimum * deltaY)
        let clippedEnd = NormalizedPoint(start.x + maximum * deltaX, start.y + maximum * deltaY)
        return (clippedStart, clippedEnd)
    }

    private static func clipTest(
        _ denominator: Double,
        _ numerator: Double,
        _ minimum: inout Double,
        _ maximum: inout Double
    ) -> Bool {
        if abs(denominator) <= epsilon {
            return numerator >= 0
        }

        let ratio = numerator / denominator
        if denominator < 0 {
            if ratio > maximum { return false }
            if ratio > minimum { minimum = ratio }
        } else {
            if ratio < minimum { return false }
            if ratio < maximum { maximum = ratio }
        }
        return true
    }

    private static func mapToCrop(_ point: NormalizedPoint, _ crop: NormalizedRect) -> NormalizedPoint {
        NormalizedPoint(
            min(max((point.x - crop.left) / crop.width, 0), 1),
            min(max((point.y - crop.top) / crop.height, 0), 1))
    }

    private static func polylineLength(_ points: [NormalizedPoint]) -> Double {
        var length = 0.0
        for index in stride(from: 1, to: points.count, by: 1) {
            length += distance(points[index - 1], points[index])
        }
        return length
    }

    private static func distance(_ first: NormalizedPoint, _ second: NormalizedPoint) -> Double {
        let x = second.x - first.x
        let y = second.y - first.y
        return (x * x + y * y).squareRoot()
    }

    private static func validateCrop(
        _ crop: NormalizedRect,
        sourcePath: String,
        pixelWidth: Int,
        pixelHeight: Int
    ) throws {
        if !crop.x.isFinite || !crop.y.isFinite || !crop.width.isFinite || !crop.height.isFinite
            || crop.x < 0 || crop.y < 0 || crop.width <= 0 || crop.height <= 0
            || crop.right > 1 || crop.bottom > 1
        {
            throw SnapikError.argumentOutOfRange(
                "crop: Crop bounds must be a positive rectangle within normalized [0, 1] image space.")
        }

        if sourcePath.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            // Port of `ArgumentException.ThrowIfNullOrWhiteSpace`: `sourcePath` is a non-optional
            // Swift `String`, so only the empty/whitespace case (ArgumentException) is reachable
            // here; the null case (ArgumentNullException) cannot occur.
            throw SnapikError.argument("sourcePath must not be empty or whitespace.")
        }
        if !RelativePathValidation.isRelativeAndSafe(sourcePath) {
            throw SnapikError.argument(
                "sourcePath: Cropped source image path must remain relative to the session directory.")
        }
        if pixelWidth <= 0 {
            throw SnapikError.argumentOutOfRange("pixelWidth")
        }
        if pixelHeight <= 0 {
            throw SnapikError.argumentOutOfRange("pixelHeight")
        }
    }
}
